docs/DifferentiableTypes.md

```diff
@@ -0,0 +1,289 @@
+    public typealias TangentVector = Tensor
+    public typealias CotangentVector = Tensor
+    public typealias AllDifferentiableVariables = Tensor
+}
+```
+
+## Compiler-synthesized implementations
+
+As shown above, the compiler automatically synthesizes implementations of `Differentiable` requirements for struct types.
+
+Here are the current conditions for synthesis:
+* The type must declare a conformance to `Differentiable`, either on the type declaration or on an extension.
+* The conforming type must be a struct type.
+* All stored properties of the conforming type must either conform to `Differentiable` or be marked with the `@noDerivative` attribute.
+
+The synthesis behavior is explained below.
+
+### Associated type synthesis
+
+Here are the synthesis rules for the three `Differentiable` associated types: `TangentVector`, `CotangentVector`, and `AllDifferentiableVariables`.
+
+Let "differentiation properties" refer to all stored properties of the conforming type that are not marked with `@noDerivative`. These stored properties are guaranteed by the synthesis condition to all conform to `Differentiable`.
+
+The synthesis rules are:
+* Set associated types to `Self`, if possible.
+  * If the conforming type conforms to `AdditiveArithmetic`, and no `@noDerivative` stored properties exist, and all stored properties satisfy `Self == Self.TangentVector == Self.CotangentVector == Self.AllDifferentiableVariables`, then all associated types can be set to typealiases of `Self`.
+* Synthesize a single `AllDifferentiableVariables` member struct. Set `TangentVector` and `CotangentVector` to `AllDifferentiableVariables` if possible; otherwise synthesize more member structs.
+  * Regarding member struct synthesis: for each "differentiation property" in the conforming type, a corresponding stored property is synthesized in the member structs, with type equal to the property’s associated type.
+  * `TangentVector` and `CotangentVector` can be set to `AllDifferentiableVariables` if all differentiation properties satisfy `Self.TangentVector == Self.CotangentVector == Self.AllDifferentiableVariables`.
+
+A memberwise initializer is synthesized for the conforming type itself, in addition to all associated structs. This is important for differentiating struct properties accesses and synthesizing manifold operation requirements.
+
+### Synthesis for other requirements
+
+`var allDifferentiableVariables: AllDifferentiableVariables` is synthesized as a computed property that mirrors the differentiation properties of the conforming type.
+
+It is always synthesized with a getter.
+It is synthesized with a setter only when all differentiation properties are mutable and themselves all have mutable `allDifferentiableVariables` properties.
+
+Examples:
+```swift
+// Example when `AllDifferentiableVariables == Self`.
+var allDifferentiableVariables: AllDifferentiableVariables {
+    get { return self }
```

Comment:
Speaking of this, we should totally change derived conformances to derive `_read` and `_modify` accessors.